### PR TITLE
new quick2vo target: like vio2vo, but smarter

### PR DESCRIPTION
**Pull request by RalfJung**
This adds a new target to do both `make quick` and `make vio2vo`, and to only `vio2vo` those files where the `vio` is newer than the `vo`.

For now, I did not touch the existing target to (a) keep compatibility -- I have no idea why this was even split into two different goals, so I don't feel like I can make the decision to remove that split -- and (b) because this is not working perfectly due to <https://coq.inria.fr/bugs/show_bug.cgi?id=5223>. That problem in principle also exists with the existing target, but there one can work around it my deleting all `*.vo` files -- due to how naive `vio2vo` is, this dosn't actually cost any time. It re-does all the work all the time anyway.

Btw, I have no idea which bugreport that comment is referring to... or maybe this is not my name misspelled, but referring to someone else?

I think a better approach, while still mostly keeping compatibility, would be to make `vio2vo` smart about timestamps as well, and to have `quick2vo` just call `quick` and then `vio2vo` -- but again, I didn't know why things work the way they work, so I didn't want to change them.

---

Recursively calling `make` is not nice. The problem with adding the `.vio` files as dependencies like for `make quick` is that one would have to do `make quick2vo -j8 J=8` -- this looks somewhat silly. The right fix for this is for Coq's `-schedule-vio2vo` to gain support for talking with make's jobserver, so that its parallelism is controlled by make.

### CODE REVIEW

**Comment by RalfJung** [comment]
There is another problem that this does not fix: `make quick2vo` or `make vio2vo` followed by e.g. `make validate` will re-compile many files because the `2vo` targets do not generate the `.vo` files in the right order, so `make` considers many of them to be outdated.

Maybe this should be an option to `coq_makefile`: Generating one of two possible makefiles.
* The standard makefile is just like now, but there is no `vio2vo`.
* The "2vo" makefile emits totally different dependencies. `.vo` files depend on imported `.vio` files (not `.vo` like they do now); `make` per default does something like `quick2vo`, and so do all commands that depend on .vo files (e.g. `make validate`).

Or maybe the makefile can actually be flexible enough to change the entire dependency structure based on a parameter (`make QUICK=1`, `make validate QUICK=1`).

**Review by gares — CHANGES_REQUESTED** [review]
The code is good, but I'd like to have a test in `test-suite/coq-makefile/`

**Comment by gares** [comment]
Another option is to have coqtop forge the date of the files in vio2vo. E.g. if each .vo gets the date of the corresponding `.vio` + 1 second, then make should be happy.  I've no clue if I can forge the dates of a file from ocaml.

**Comment by gares** [comment]
Looks like `Unix.utimes` does the trick...

I guess one could `Unix.stat` the input file `f` https://github.com/coq/coq/blob/master/toplevel/vernac.ml#L331 and add an argument to save_library_raw (the mtime) and have such function call `Unix.utimes` on the output file.

**Comment by RalfJung** [comment]
> I'd like to have a test in test-suite/coq-makefile/

What should that test do? I looked at some of them, but most seem related to plugins (which I have very little understanding of), and `arg` seems to just call `make` and be done?

**Comment by maximedenes** [comment]
@gares can you explain what you had in mind? Thanks!

**Comment by gares** [comment]
Yep, I just want the test to run the new make target and eventually check that .vo files are there once it terminates.

**Comment by RalfJung** [comment]
I tried adding a test, but can't even seem to get existing tests to tun. I did `./configure -local`, added the `bin` directory to my PATH, and then did `./run.sh` in `test-suite/coq-makefile/validate1`, but that fails:
```
CAMLOPT -pack -o src/test_plugin.cmx
ocamlfind: Package `camlp5.gramlib' not found
Makefile:633: recipe for target 'src/test_plugin.cmx' failed
make[1]: *** [src/test_plugin.cmx] Error 2
make[1]: Leaving directory '/home/r/src/coq/coq/test-suite/coq-makefile/quick2vo'
Makefile:385: recipe for target 'quick2vo' failed
make: *** [quick2vo] Error 2
```

There's no README anywhere in the test-suite as far as I can tell, so I have no idea how to even develop a test here. Also, on my first attempt, the test just deleted itself when I ran it due to a `git clean -dfx` in `template/init.sh`. That's certainly NOT what I expected it to do...

**Comment by Zimmi48** [comment]
> There's no README anywhere in the test-suite as far as I can tell, so I have no idea how to even develop a test here. Also, on my first attempt, the test just deleted itself when I ran it due to a git clean -dfx in template/init.sh. That's certainly NOT what I expected it to do...

I hate it that the test-suite contains such things. First it should not depend on `git`, second it's a pretty dangerous thing to put in a script. Please someone fix that!

**Comment by RalfJung** [comment]
Ping @gares -- I am waiting for some guidance for how to write and debug coq_makefile tests.

**Comment by gares** [comment]
I did miss this ping, I'm sorry.
In the meanwhile the Makefile has been fixes not to call git clean...

Copying an existing test is the best way to go.
To run a test I do something like `make -C test-suite coq-makefile/validate1.log` (it is not documented, but not specific to the coq-makefile tests either).

**Comment by RalfJung** [comment]
Blocked by #6691. Looks like `-schedule-vio2vo` is broken in master?

**Comment by RalfJung** [comment]
All right, I added tests. Let's see what the CI says.

File: tools/CoqMakefile.in [hunk 1]
```diff
@@ -381,7 +381,7 @@ bytefiles: $(CMOFILES) $(CMAFILES)
 optfiles: $(if $(DO_NATDYNLINK),$(CMXSFILES))
 .PHONY: optfiles
 
-# FIXME, see Ralph's bugreport
+# FIXME, see Ralf's bugreport
```
Comment (RalfJung) [inline tools/CoqMakefile.in]: Btw, does anyone know which bugreport is meant here? This line was already present when coq_makefile2 was first committed by @gares. It doesn't seem to be present before that though.

**Comment by RalfJung** [comment]
Btw, do I have to register the new tests anywhere or are all directories picked up automatically?

**Comment by RalfJung** [comment]
> COQC -quick theories/test.v
VIO2VO: -n theories/test.vio 
coqc: -n: no such file or directory

Now that is strange, somehow there is a `-n` creeping into the list of file names.

**Comment by RalfJung** [comment]
Equations CI failure doesn't look like it could be introduced by this PR:
```
File "src/equations_common.ml", line 672, characters 2-9:
Error: This function has type
         Evd.evar_map -> EConstr.constr -> EConstr.constr
       It is applied to too many arguments; maybe you forgot a `;'.
```

Windows OTOH seems legit -- it just doesn't support vio2vo?
```
make[1]: Leaving directory '/cygdrive/c/projects/coq/test-suite/coq-makefile/vio2vo/_test'
make[1]: Entering directory '/cygdrive/c/projects/coq/test-suite/coq-makefile/vio2vo/_test'
"/cygdrive/c/projects/coq/bin/coqc"  -q  -I src  -R src test -R theories test  \
	-schedule-vio2vo 2 theories/test.vio
Error:
Anomaly "Uncaught exception Invalid_argument("Unix.wait not implemented")."
Please report at http://coq.inria.fr/bugs/.
```

**Comment by RalfJung** [comment]
@gares What should I do? Disable the test on Windows or wait until `schedule-vio2vo` is fixed on Windows?

**Comment by Zimmi48** [comment]
Clearly the first one. You should find examples of already disabled tests for Windows in the test-suite...

**Comment by RalfJung** [comment]
Done. I also rebased the PR to hopefully fix the equations test.

File: test-suite/coq-makefile/quick2vo/run.sh [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+a=`uname`
+
+. ../template/init.sh
+
+coq_makefile -f _CoqProject -o Makefile
+# vio2vo is broken on Windows (#6720)
+if [ "$a" = "Darwin" -o "$a" = "Linux" ]; then
```
Comment (Zimmi48) [inline test-suite/coq-makefile/quick2vo/run.sh]: You should print a warning if the test is not run.

Comment (RalfJung) [reply test-suite/coq-makefile/quick2vo/run.sh]: I copied this from `misc/deps-utf8` which also does not. How should the warning look like?

Comment (Zimmi48) [reply test-suite/coq-makefile/quick2vo/run.sh]: OK forget about this, there are many tests that are not run without printing a warning and this should be addressed globally.

**Comment by RalfJung** [comment]
All right, CI is all green finally. :) @gares ready to merge?

**Review by gares — APPROVED** [review]

File: tools/CoqMakefile.in (hunk 1 above)
Comment (maximedenes) [reply tools/CoqMakefile.in]: @gares ping

**Comment by maximedenes** [comment]
@RalfJung I'll squash a bit, push to your branch, and merge, if you don't mind.

**Comment by Zimmi48** [comment]
FTR I had a spurious failure (locally) with one of these new tests. The log file contained:

```
Error:
Anomaly
"Uncaught exception Failure("The spawned process did not connect back in 2.0s")."
Please report at http://coq.inria.fr/bugs/.
```

It is similar to #5983.

**Pull request merged**